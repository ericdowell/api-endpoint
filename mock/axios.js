export function BasicMock(options) {
  return new Promise((resolve, reject) => {
    if (options.errors) {
      return reject({
        config: {},
        response: {
          data: {
            errors: options.errors
          },
          status: options.status || 422
        }
      })
    }
    return resolve({
      config: {},
      data: { foo: 'bar' }
    })
  })
}
