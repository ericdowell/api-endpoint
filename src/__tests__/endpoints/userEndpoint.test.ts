import { UserEndpoint } from '../..'
import axios from 'axios'
import { BasicMock } from '../mock/axios'

jest.spyOn(axios, 'request').mockImplementation(BasicMock)

const callEndpoint = (instance: UserEndpoint): {[key: string]: Function} => {
  return {
    show: instance.show.bind(instance),
    update: instance.update.bind(instance),
    register: instance.register.bind(instance),
    resendEmailVerification: instance.resendEmailVerification.bind(instance),
    requestPasswordReset: instance.requestPasswordReset.bind(instance),
    changePassword: instance.changePassword.bind(instance),
  }
}

describe(UserEndpoint.name, (): void => {
  it.each([
    ['show', 'get', [123, { include: ['name'] }, false]],
    ['update', 'put', [4321, { name: 'John', email: 'email' }]],
    ['register', 'post', ['email', 'email', 'password', 'password']],
    ['register', 'post', ['email', 'email', 'password', 'password', { name: 'John' }, false]],
    ['resendEmailVerification', 'post', []],
    ['requestPasswordReset', 'post', ['email']],
    ['changePassword', 'put', ['current', 'password', 'password']],
  ])('the %s calls %s parent method', async(method: string, calls, params): Promise<void> => {
    const endpoint = new UserEndpoint()
    // eslint-disable-next-line @typescript-eslint/ban-ts-ignore
    // @ts-ignore
    const methodCalled = jest.spyOn(endpoint, calls).mockImplementation((): any => true)
    await callEndpoint(endpoint)[method](...params)
    expect(methodCalled).toHaveBeenCalledTimes(1)
    expect(methodCalled.mock.calls[0]).toMatchSnapshot()
  })
})
