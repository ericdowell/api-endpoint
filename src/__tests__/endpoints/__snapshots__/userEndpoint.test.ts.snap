// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserEndpoint the changePassword calls put parent method 1`] = `
Array [
  "/password/change",
  Object {
    "data": Object {
      "current_password": "current",
      "password": "password",
      "password_confirmation": "password",
    },
  },
]
`;

exports[`UserEndpoint the register calls post parent method 1`] = `
Array [
  "/register",
  Object {
    "data": Object {
      "email": "email",
      "email_confirmation": "email",
      "password": "password",
      "password_confirmation": "password",
      "remember": true,
    },
  },
]
`;

exports[`UserEndpoint the register calls post parent method 2`] = `
Array [
  "/register",
  Object {
    "data": Object {
      "email": "email",
      "email_confirmation": "email",
      "name": "John",
      "password": "password",
      "password_confirmation": "password",
      "remember": false,
    },
  },
]
`;

exports[`UserEndpoint the requestPasswordReset calls post parent method 1`] = `
Array [
  "/password/email",
  Object {
    "data": Object {
      "email": "email",
    },
  },
]
`;

exports[`UserEndpoint the resendEmailVerification calls post parent method 1`] = `
Array [
  "/email/resend",
]
`;

exports[`UserEndpoint the show calls get parent method 1`] = `
Array [
  "/user/123",
  Object {
    "params": Object {
      "include": Array [
        "name",
      ],
    },
  },
]
`;

exports[`UserEndpoint the update calls put parent method 1`] = `
Array [
  "/user/4321",
  Object {
    "data": Object {
      "email": "email",
      "name": "John",
    },
  },
]
`;
